[Affiliate] Wire up flow updated

ramirotw opened the pull request:
# Summary

Updates #944 

Implements latest changes, mainly the IPFS upload. This PR includes changes from https://github.com/gnosis/cowswap/pull/1393, once that PR is merged I'll rebase this one and resolve any conflicts.

![image](https://user-images.githubusercontent.com/3328006/132336672-f3b43d30-eb27-4068-933c-8cb827153189.png)

# To Test
Testing this PR is not possible at the moment as it depends on environment variables for the Pignata service. Only local testing is possible.

alfetopito commented on src/custom/components/AffiliateStatusCheck/index.tsx:
```diff
@@ -0,0 +1,111 @@
+    try {
+      const appDataHash = await uploadMetadataDocToIpfs(generateReferralMetadataDoc(referralAddress))
+
+      appDispatch(updateAppDataHash(appDataHash))
```
Nitpick: why not create a hook for this as well?

alfetopito reviewed (commented):
Tiny comment for now. Will test properly once IPFS PR is merged.

Create another issue for updating metadata doc, actually makes sense to do it separately https://github.com/gnosis/cowswap/issues/1412

anxolin commented on src/custom/utils/trade.ts:
```diff
@@ -147,3 +150,9 @@ export async function sendOrderCancellation(params: OrderCancellationParams): Pr
 }
+
+export async function hasTrades(chainId: ChainId, address: string): Promise<boolean> {
+  const trades = await getTrades(chainId, address)
```
I know there's no pagination in the endpoint yet, however we know there will be
Why not making `getTrades` receive some params with `chainId`, `address` and an optional `limit` and `offset`.

Then, you would use `limit = 1` here.
Also, you would include the limit and offset params in the endpoint. It won't take effect for now, but nice to have it already since we know the "signature" it will have.

alfetopito replied:
Ah, that's why the change. Good having it now, doesn't hurt.

anxolin commented on src/custom/utils/metadata.ts:
```diff
@@ -49,3 +59,17 @@ export function generateAppDataDoc(metadata: MetadataDoc = {}): AppDataDoc {
+    return `0x${Buffer.from(digest).toString('hex')}`
+  } catch (err) {
+    console.error('There was an error uploading metadata doc to IPFS', err)
+    throw err.message ? err : new Error(err)
```
Why handling for rethrowing in this case?

anxolin commented on src/custom/utils/metadata.ts:
```diff
@@ -23,6 +28,11 @@ export type AppDataDoc = {
 
+const pinata = pinataSDK(
+  process.env.REACT_APP_PINATA_API_KEY as string,
+  process.env.REACT_APP_PINATA_SECRET_API_KEY as string
```
With the `as` u are saying that these params are provided always? 
Shouldn't you have a conditional logic? even if it's just for throwing an error if they are not provided.

anxolin commented on src/custom/utils/metadata.ts:
```diff
@@ -1,3 +1,8 @@
+import IPFS from 'ipfs-core'
+import CID from 'cids'
+import multihashes from 'multihashes'
+import pinataSDK from '@pinata/sdk'
```
We have a dir `custom/api` with API integrations. Maybe this fits more there? At least most of `uploadMetadataDocToIpfs` code

alfetopito replied:
FYI, the IPFS related changes are from https://github.com/gnosis/cowswap/pull/1393

I recommend you comment there :)

anxolin replied:
Done :)

anxolin reviewed (commented):
Is the base PR correct? I see a few changes that doesn't look they are related. 

I didn't do a functional test yet, happy to do so when there's some test instructions, probably will come in a future PR?

Anyways, nice PR!

ramirotw commented on src/custom/api/gnosisProtocol/index.ts:
```diff
@@ -10,7 +10,6 @@ const useMock = process.env.REACT_APP_MOCK === 'true'
 export const {
   // functions that have a mock
-  getAppDataDoc,
   uploadAppDataDoc,
```
@alfetopito does this `uploadAppDataDoc` still make any sense? or should I remove it on this PR to wrap up all the Affiliate related changes?

alfetopito replied:
You are right, it no longer applies as it'll be uploaded directly to IPFS.
Please remove it

ramirotw commented:
FYI, some of the PRs that were needed for this wire up were merged so I did a rebase and also changed the base branch to point to Matias' [IPFS upload PR](https://github.com/gnosis/cowswap/pull/1393). Now the changes should be clearer for reviewing.

alfetopito commented on src/custom/api/gnosisProtocol/api.ts:
```diff
@@ -255,48 +265,43 @@ export async function getOrder(chainId: ChainId, orderId: string): Promise<Order
+
+export async function getTrades(params: GetTradesParams): Promise<TradeMetaData[]> {
+  const { chainId, owner, limit, offset } = params
+  const qsParams = stringify({ owner, limit, offset })
```
You are a bit ahead as the current `/trades` endpoint do not have pagination.
Also, pagination like this will be present on the new trades endpoint (`/api/v1/account/{owner}/trades`), similar to the new orders endpoint: https://protocol-rinkeby.dev.gnosisdev.com/api/#/default/get_api_v1_account__owner__orders

Although, that's not available yet.

All this is to say this query might not work as you expect and return all trades instead.

But that is ok, I believe the default behaviour is to ignore unknown parameters.

No need to change anything, just FYI

anxolin replied:
Yep, kind of nice that we are already prepared, I would also expect them to ignore them if it's not implemented, and we will see it working as soon as they implement it, so not bad to leave it imo

alfetopito commented:
I've tested the flow and it's working fine, except for the IPFS upload part.
It's missing a setting on CI I've addressed on https://github.com/gnosis/cowswap/pull/1470

Once that's merged it should be all good

alfetopito commented:
FYI, waiting on Anxo to review. He's on holiday today and will be back tomorrow

anxolin commented on src/custom/components/AffiliateStatusCheck/index.tsx:
```diff
@@ -0,0 +1,108 @@
+
+const STATUS_TO_MESSAGE_MAPPING: Record<AffiliateStatus, string> = {
+  NOT_CONNECTED: 'Please connect your wallet to participate',
+  OWN_LINK: 'You followed your own referral link',
```
nit: I would phrase it as:
"Your affiliate link works and It's ready to be shared!"

I find it has a more positive note (just not to feel they did sth wrong by following), and reassures the user.

anxolin commented on src/custom/components/AffiliateStatusCheck/index.tsx:
```diff
@@ -0,0 +1,108 @@
+  const [affiliateState, setAffiliateState] = useState<AffiliateStatus | null>()
+  const [error, setError] = useState('')
+
+  // TODO: retry on error?
```
Not for this PR, but you can add to an issue as a checkbox, or create an issue for this.

There's an util u can use: `utils/retry`

ramirotw replied:
Great, I can wrap the `hasTrades` and the `uploadReferralDocAndSetDataHash` calls individually with that retry helper. It's a small change that if it makes sense I can add it on this PR.

Should we use a common a common retry options setup such as: 

```js
{
  n: 3, 
  minWait: 1000, 
  maxWait: 3000
}
```
?

anxolin replied:
That works!

anxolin reviewed (commented):
Approved code wise, but I haven't been able to test this. I'm not sure what I'm doing wrong, but https://pr1405--gpswapui.review.gnosisdev.com/#/profile shows me NOT FOUND

anxolin reviewed (commented):
Looks mainly good!

I just saw something strange with following your own link. It doesn't show the message unless you are in the swap page.

1. Follow your own link
2. It redirects to the profile page, but doesn't show any message
3. Only if I go back to trade page shows that message

ramirotw commented:
> Looks mainly good!
> 
> I just saw something strange with following your own link. It doesn't show the message unless you are in the swap page.
> 
> 1. Follow your own link
> 2. It redirects to the profile page, but doesn't show any message
> 3. Only if I go back to trade page shows that message

Fixed!

> In addition to Anxo comment, I think it would be nice if a user will have a possibility to close this banner. Currently, close icon on banner does not work in Win10+Chrome.

Also fixed!
